mqtt-client-C-paho/emqx_file_transfer.c

```diff
@@ -0,0 +1,290 @@
+/*
+ * This is an example of how to use the MQTT client library to send a file to
+ * the broker using EMQX's the file transfer extension.
+ *
+ * The EMQX file transfer extension documentation is available here:
+ * https://www.emqx.io/docs/en/v5/file-transfer/introduction.html
+ *
+ * This example reads a file from the file system and publishes it to the
+ * broker (the file transfer feature has to be enabled in the EMQX
+ * configuration). The user can specify the file path, file id, file name etc
+ * as command line parameters. Run the program with the --help flag to see the
+ * list of options.
+ *
+ * Change the DEBUG macro to 1 to see debug messages.
+ */
+
+
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <time.h>
+
+#include <MQTTClient.h>
+
+#define CLIENTID    "c-client"
+#define TIMEOUT     10000L
+#define DEBUG       0
+
+int send_file(MQTTClient client,
+              char *file_path,
+              char *file_id,
+              char *file_name,
+              unsigned long expire_time_s_since_epoch,
+              unsigned long segments_ttl_seconds) {
+    FILE *fp = fopen(file_path, "rb");
+    int rc;
+    int qos = 1;
+    const size_t buf_size = 2048;
+    if (fp == NULL) {
+        printf("Failed to open file %s\n", file_path);
+        return -1;
+    }
+    // Get file size
+    fseek(fp, 0L, SEEK_END);
+    long file_size = ftell(fp);
+    fseek(fp, 0L, SEEK_SET);
+    // Create payload for initial message 
+    char payload[buf_size];
+    rc = snprintf(payload,
+            buf_size,
+            "{\n"
+            "  \"name\": \"%s\",\n"
+            "  \"size\": %ld,\n"
+            "  \"expire_at\": %ld,\n"
+            "  \"segments_ttl\": %ld,\n"
+            "  \"user_data\": {}\n"
+            "}",
+            file_name,
+            file_size,
+            expire_time_s_since_epoch,
+            segments_ttl_seconds);
+    if (rc < 0 || rc >= buf_size) {
+        printf("Failed to create payload for initial message\n");
+        return -1;
+    }
+    // Create topic of the form $file/{file_id}/init for initial message
+    char topic[buf_size];
+    MQTTClient_deliveryToken token;
+    rc = snprintf(topic, buf_size, "$file/%s/init", file_id);
+    if (rc < 0 || rc >= buf_size) {
+        printf("Failed to create topic for initial message\n");
+        return -1;
+    }
+    // Publish initial message
+    if (DEBUG) {
+        printf("Publishing initial message to topic %s\n", topic);
+        printf("Payload: %s\n", payload);
+    }
+    rc = MQTTClient_publish(client, topic, strlen(payload), payload, 1, 0, &token);
+    if (rc != MQTTCLIENT_SUCCESS) {
+        printf("Failed to publish message, return code %d\n", rc);
+        return -1;
+    }
+    rc = MQTTClient_waitForCompletion(client, token, TIMEOUT);
+    if (rc != MQTTCLIENT_SUCCESS) {
```

Comment:
Q: As per [documentation](https://eclipse.github.io/paho.mqtt.c/MQTTClient/html/_m_q_t_t_client_8h.html#a83807ec81fe8c3941e368ab329d43067) this doesn't (evidently) give us a PUBACK Reason code here? I wonder what `rc` we'll get here if EMQX send a PUBACK with non-zero Reason code.

Reply:
This is a very good question! I will try to scan the documentation and try to find out how the get the PUBACK reason code.

Reply:
I did some digging and it seems like the only way to get the PUBACK Reason code is to register a publication handler https://www.eclipse.org/paho/files/mqttdoc/MQTTClient/html/_m_q_t_t_client_8h.html#a9f13911351a3de6b1ebdabd4cb4116ba . However, this handler is only used for version 5 of the MQTT protocol if I understand the documentation correctly. The default MQTT version for the POHA C library is 3.X and when I explicitly set the version to 5 in the connect options the client fails to connect with a strange exit code. So at the moment I'm not sure how to proceed with this (maybe turning on some debugging option might reveal why it fails to connect).

By the way the PUBACK Reason seems to be a MQTT version 5 only feature and the documentation does not seem to mention that (https://www.emqx.io/docs/en/v5/file-transfer/client.html#the-common-flow). Maybe this is something that we should fix?

Reply:
Added a commit with a TODO comment in the code with info about this issue. Will merge this now.